### PR TITLE
Initial version of the dfavm

### PR DESCRIPTION
**Pull request by sfstewman**
dfavm is a bytecode based way to run DFA matching.  This commit provides:

1. Initial cut at the dfavm "compiler", dfavm bytecode, and dfavm vm to run the bytecode.
2. `retest`, a program to test dfavm matching of regular expressions.
3. Two utility programs to adapt the PCRE test suites to something `retest` can run

### CODE REVIEW

**Comment by sfstewman** [comment]
The dfavm compiler does some amount of basic block reordering, but further optimization will come in a later PR.

File: format_pcre_tests.sh [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#! /bin/sh
```
Comment (katef) [inline format_pcre_tests.sh]: You can also do like
```
#!/usr/bin/jq -srf
```

File: src/fsm/main.c [hunk 2]
```diff
@@ -312,6 +314,32 @@ xopen(const char *s)
...
+	struct prng_state prng;
+
+	memset(&prng, 0, sizeof prng);
+	prng_seed(&prng, seed);
+
+	fsm_generate_words_to_file(fsm, &params, &prng, num_words, f);
+}
+#endif /* 0 */
+
+
```
Comment (katef) [inline src/fsm/main.c]: What's going on here? Is `fsm_generate_words_to_file()` just not present yet?

File: src/libfsm/dfavm.c [hunk 3]
```diff
@@ -0,0 +1,1692 @@
...
+#include <string.h>
+
+#include <fsm/fsm.h>
+#include <fsm/vm.h>
+#include "internal.h"
+#include "print/ir.h"
+
+#define DEBUG_ENCODING     0
+#define DEBUG_VM_EXECUTION 0
+#define DEBUG_VM_OPCODES   0
```
Comment (katef) [inline src/libfsm/dfavm.c]: I'm not sure about the encoding and execution, but I do think `DEBUG_VM_OPCODES` should be exposed to both `re(1)` and `fsm(1)` as a `-l ...` output language option. (You can treat the `fsm_print` function pointer essentially as an enum for that)

File: src/libfsm/dfavm.c [hunk 4]
```diff
@@ -0,0 +1,1692 @@
...
+	case VM_OP_BRANCH:
+		{
+			char dst;
+			switch (op->u.br.dest) {
+			case VM_DEST_NONE:  dst = 'Z'; break;
+			case VM_DEST_SHORT: dst = 'S'; break;
+			case VM_DEST_NEAR:  dst = 'N'; break;
+			case VM_DEST_FAR:   dst = 'F'; break;
+			default:            dst = '!'; break;
+			}
```
Comment (katef) [inline src/libfsm/dfavm.c]: I do agree that the near/far jumps are probably overkill, but not strongly :)

File: src/libfsm/dfavm.c [hunk 5]
```diff
@@ -0,0 +1,1692 @@
...
+fsm_vm_match_buffer(const struct fsm_dfavm *vm, const char *buf, size_t n)
+{
+	struct vm_state st = state_init;
+	enum dfavm_state result;
+
+	vm_match(vm, &st, buf, n);
+	result = vm_end(vm, &st);
+
+	return result == VM_SUCCESS;
+}
```
Comment (katef) [inline src/libfsm/dfavm.c]: Maybe not on this initial PR, but it'd probably be sensible to split this into several files, for compilation, execution, etc.

File: src/retest/main.c [hunk 6]
```diff
@@ -0,0 +1,793 @@
...
+	if (cpy == NULL) {
+		if (err != NULL) {
+			*err = -1;
+		}
+		return NULL;
+	}
+
+	memcpy(cpy, s, len+1);
+	return cpy;
+}
```
Comment (katef) [inline src/retest/main.c]: There's an `xstrdup()` in adt/xalloc.c if you want it

File: format_pcre_tests.sh (hunk 1 above)
Comment (sfstewman) [reply format_pcre_tests.sh]: Nice!

Semi-relatedly, I'm thinking of scrapping this script and either adapting the `extract_pcre_tests.py` script to generate `retest` format directly or writing a small C program to do this (removing the python dependency).

File: src/fsm/main.c (hunk 2 above)
Comment (sfstewman) [reply src/fsm/main.c]: Oh, whoops!  That was part of my on-hold project to generate text directly from a DFA.  Must have slipped in there...

File: src/libfsm/dfavm.c (hunk 3 above)
Comment (sfstewman) [reply src/libfsm/dfavm.c]: That makes sense.

I'm thinking we probably want to split things so there's a high-level "IR-ish" set of opcodes and a bytecode encoding.  The opcodes would be useful for optimization and would be the basis for any generated bytecode and assembly.

In this scheme, you'd want `fsm_print` functions for both the opcodes and bytecodes.

File: src/libfsm/dfavm.c (hunk 4 above)
Comment (sfstewman) [reply src/libfsm/dfavm.c]: As soon as I have a performance test suite, I'm going to try both 16-bit and 32-bit fixed width bytecodes for the VM.

File: src/libfsm/dfavm.c (hunk 5 above)
Comment (sfstewman) [reply src/libfsm/dfavm.c]: Yes!  This file is getting too large.

**Pull request merged**